cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/config_renderer.py

```diff
@@ -0,0 +1,177 @@
+import logging
+from typing import Dict
+
+from .config_utils import (
+    _get_efa_settings,
+    _get_instance_types,
+    _get_min_gpu_count_and_type,
+    _get_min_vcpus,
+    _get_real_memory,
+)
+
+CONFIG_HEADER = "# This file is automatically generated by pcluster"
+log = logging.getLogger()
+
+
+class ComputeResourceRenderer:
+    """Renders a PCluster ComputeResource Config as a Slurm NodeName config, gres config or node_set element."""
+
+    def __init__(self, queue_name: str, compute_resource_config: Dict, no_gpu, memory_ratio, instance_types_data: Dict):
+        self.queue_name = queue_name
+        self.has_gpu = not no_gpu
+        self.instance_types_data = instance_types_data
+        self.static_nodes = compute_resource_config["MinCount"]
+        self.dynamic_nodes = compute_resource_config["MaxCount"] - self.static_nodes
+        self.name = compute_resource_config["Name"]
+        self.disable_simultaneous_multithreading = compute_resource_config["DisableSimultaneousMultithreading"]
+        self.spot_price = compute_resource_config.get("SpotPrice", None)
+        self.instance_types = _get_instance_types(compute_resource_config)
+        self.real_memory = _get_real_memory(
+            compute_resource_config, self.instance_types, instance_types_data, memory_ratio
+        )
+        self.EFA_enabled, self.EFA_gdr_support = _get_efa_settings(compute_resource_config)
+        self.instance_types = _get_instance_types(compute_resource_config)
+        self.vcpus_count, self.threads_per_core = _get_min_vcpus(self.instance_types, instance_types_data)
+        self.gpu_count, self.gpu_type = _get_min_gpu_count_and_type(self.instance_types, instance_types_data, log)
+
+    def static_node_name(self):
+        """Render the NodeName section for static nodes."""
+        return self._node_name("st", self.static_nodes)
+
+    def dynamic_node_name(self):
+        """Render the NodeName section for dynamic nodes."""
+        return self._node_name("dy", self.dynamic_nodes)
+
+    def vcpus(self) -> int:
+        """Return the number of vcpus according to disable_hyperthreading and instance features."""
+        return (
+            self.vcpus_count
+            if not self.disable_simultaneous_multithreading
+            else (self.vcpus_count // self.threads_per_core)
+        )
+
+    def gpus(self) -> dict:
+        """Return the number of GPUs and type for the compute resource."""
+        return {"count": self.gpu_count, "type": self.gpu_type}
+
+    def render(self):
+        """Launch the rendering process."""
+        config = ""
+        if self.static_nodes > 0:
+            config += f"NodeName={self.static_node_name()}{ self._definitions() }\n"
+        if self.dynamic_nodes > 0:
+            config += f"NodeName={self.dynamic_node_name()}{ self._definitions(dynamic=True) }\n"
+
+        return config
+
+    def as_node_set(self):
```

Comment:
Why don't we split the NodeSet as a trivial separate class? Maybe we don't need it now, but in the long run it may be cleaner.

Reply:
The reasoning behind this decision is that it would require to duplicate the list of ComputeResource objects or at least to share the same references with the Queue class.
And the advantage would be to move a single method.
I think we aren't there yet.
If/when there will be a richer behavior than just formatting a string we will add it.

Reply:
Done